implementations/rust/ockam/ockam_core/src/cancel.rs

```diff
@@ -0,0 +1,67 @@
+use crate::{Address, LocalMessage, Message, NodeContext, Result, Routed};
+
+/// A message wraper type that allows users to cancel message receival
+///
+/// A worker can block in place to wait for a message.  If the next
+/// message is not the desired type, it can be cancelled which
+/// re-queues it into the mailbox.
+pub struct Cancel<'ctx, C, M> {
+    inner: M,
+    local_msg: LocalMessage,
+    addr: Address,
+    ctx: &'ctx C,
+}
+
+impl<'ctx, C: NodeContext, M: Message> Cancel<'ctx, C, M> {
+    /// Create a new `Cancel`. This should only be used inside node
+    /// implementations.
+    #[doc(hidden)]
+    pub fn new(inner: M, local_msg: LocalMessage, addr: Address, ctx: &'ctx C) -> Self {
+        Self {
+            inner,
+            local_msg,
+            addr,
+            ctx,
+        }
+    }
+
+    /// Cancel this message
+    pub async fn cancel(self) -> Result<()> {
+        self.ctx.forward(self.local_msg).await
```

Comment:
Off-topic: if cancelling a message became that simple, do we even need this structure?

Reply:
I think it's still valuable. Not because of the complexity involved in this function, but the complexity on the user's part to understand the internals of the runtime enough to connect that `forward(...)` re-queues the message.

Reply:
@spacekookie how about moving that to Routed<>?